### PR TITLE
grpc exception handling

### PR DESCRIPTION
**Pull request by anjeyy**
**Work in progress** - Feedback appreciated

I stumbled upon a cluttered Error Handling with gRPC and i was used to the  _spring way_ how to handle Exceptions for ReST-Calls.
So i adopted a little bit from Springs `@ControllerAdvice` and created something for gRPC `@GrpcServiceAdvice`.

I'll work an in-depth example into the docs.

If you guys have any suggestions or improvements, please let me know. 

---
# Workflow
- Service is annotated with `@GrpcService`
- Create Class with annotation `@GrpcServiceAdvice`
- Annotate Methods with `@GrpcExceptionHandler`
    - add wanted Exception class inside annotation OR
    - give the annotated Method wanted Exception as parameter
    - or both of the above
- annotated Method has to return `Throwable`, since gRPC needs to call for a request `StreamObserver<?>.onError(Throwable t)`

---

# TODO

- [x] closing connection for _grpc stream_ (asyn calls) ✔️ 
    - exception is being handled, but the stream is not closed
- [x] add unit tests ✔️ 
- [x] add java docs ✔️ 
- [x] add project docs (.md files) ✔️
- [x] waiting for #471 merge ✔️ 
- [ ] review 🏁 _in progress_

EDIT: updated todos

### CODE REVIEW

**Comment by ST-DDT** [comment]
Nice, thanks for your contribution!

Could you please send me a link to the related `@ControllerAdvice` logic for comparison?
Also I'm somewhat unsure, whether I prefer this pointcut logic or grpc interceptors.

> annotated Method has to return Throwable, since gRPC needs to call for a request StreamObserver<?>.onError(Throwable t)

IMO it should also/only be allowed (and recommended) to return `io.grpc.Status` as it can easily be converted to a `RuntimeStatusException` and is more clear about the resulting behavior is. 
If it returns (or throws an exception), what will happen next? Will another handler be tried? If we return a status code then it's more obvious. Also on the client side we won't ever receive the intermediate exceptions.

What do you think?

**Comment by anjeyy** [comment]
Hi @ST-DDT,

[@ControllerAdvice](https://github.com/spring-projects/spring-framework/blob/master/spring-web/src/main/java/org/springframework/web/bind/annotation/ControllerAdvice.java) is the Annotation which binds for a `@Controller`/`@RestController`, where the response is beeing sent back.
[Here](https://github.com/spring-projects/spring-framework/blob/master/spring-web/src/main/java/org/springframework/web/method/annotation/ExceptionHandlerMethodResolver.java) is the Class `ExceptionHandlerMethodResolver` where i did look a little bit at how Spring resolves their annotation `@ExceptionHandler` on a method and maps it to the corresponding Exception.

If you look at how to [make use](https://dzone.com/articles/best-practice-for-exception-handling-in-spring-boo) of `@ControllerAdvice`, i tried to do the same, but for the annotation `@GrpcService`. Every Exception thrown inside a Class which is annotated with `@GrpcService` **and** the Class implements `io.grpc.BindableService`. This logic is expressed through Class [`GrpcExceptionAspect`](https://github.com/yidongnan/grpc-spring-boot-starter/pull/465/files#diff-66d5f456b44a7b84093e5c87e221c74b9da3a06331537601c7c662b705610348).

I strongly recommend exception handling to be familiar to spring annotations, since this is a _spring-boot-starter_ library so that developer can adopt already familiar features quite easily.
I am not that familiar with the grpc interceptors, how can we achieve a annotation driven exception handling?
Three Major things to consider
- Advice Class where Exceptions are beeing declared (to be mapped)
- declare Exceptions to be handled, when thrown
- interrupting, when exception occurs
    - also closing the stream with [`StreamObserver.onError(Throwable t)`](https://github.com/grpc/grpc-java/blob/821ec65f2ede1adaa54400e0a093597ec7958485/stub/src/main/java/io/grpc/stub/StreamObserver.java#L69)

---

About the second part with the returning throwable to call `StreamObserver.onError(...)`. When you please have a look at my [example grpc setup](https://github.com/anjeyy/grpc-sample/blob/grpc-exception-handling-generic/grpc-server/src/main/java/com/github/anjeyy/infrastructure/exception/GrpcExceptionAdvice.java), there you can call also `return status.asException();` and/or put `Metadata` inside with `status.asRuntimeException(myMetadata)`. 
This mapped Method is beeing executed in the Advice and the return type `Throwable` is then beeing sent with the `StreamObserver.onError(...)`. Similar to springs `ResponseEntity.status(NOT_FOUND).body(...)`.

How would the workflow look if the Advice is returning `Status`? I think the stream needs to be closed anyway, right? So instead of letting the developer have to think about it, the Aspect is closing the steam with the resulting error. Or do you think to leave it open to the developer?

If i am overlooking something, please let me know.

**Comment by ST-DDT** [comment]
Thanks for the link.
Could you please also send me a link to the spring-web variant of `GrpcExceptionAspect`?

> there you can call also return status.asException();

I'm aware of that. But IMO this `asException()` is unnecessary. What else could a developer intend other than this status being sent to the client? If any exception were to be sent to the client as is, then I wouldn't mind, but the exceptions are effectively reduced to a `Status` message. So it should be possible to keep the user code free of these conversions/wrappers. 
In spring I prefer not no use the `ResponseEntity` wrappers unless I actually need them.

> How would the workflow look if the Advice is returning Status? I think the stream needs to be closed anyway, right? So instead of letting the developer have to think about it, the Aspect is closing the steam with the resulting error. Or do you think to leave it open to the developer?

No, the aspect would look and behave exactly the same, except for the check, if the exception handler returned a Status or an Exception. The Status would be converted to an exception and then be passed on.

> and/or put Metadata inside with status.asRuntimeException(myMetadata).

I agree, this is a valid point to also allow exceptions to be returned.


> I strongly recommend exception handling to be familiar to spring annotations, since this is a _spring-boot-starter_ library so that developer can adopt already familiar features quite easily.
> I am not that familiar with the grpc interceptors, how can we achieve a annotation driven exception handling?
> Three Major things to consider
> 
>     * Advice Class where Exceptions are beeing declared (to be mapped)
> 
>     * declare Exceptions to be handled, when thrown
> 
>     * interrupting, when exception occurs
>       
>       * also closing the stream with [`StreamObserver.onError(Throwable t)`](https://github.com/grpc/grpc-java/blob/821ec65f2ede1adaa54400e0a093597ec7958485/stub/src/main/java/io/grpc/stub/StreamObserver.java#L69)

I completely agree to that (on the usage side that is).
But there is one point that we needs to keep in mind. grpc also has streaming calls/is implicitly (potentially) async (as you can return from the controller method, without completing the StreamObserver) so it should be possible to just "quit" processing anywhere in the code in any thread by calling `StreamObserver#onError && return` and still have the exception handling kick in.

This is similar to default spring/spring-security, where you can (to a certain degree) handle exceptions outside of your controller implementation. IIRC the web exception handling also kicks in for methods that return unresolved `Future`s.

**Comment by ST-DDT** [comment]
You probably know this already, but I thought about adding this dictionary to the docs eventually.

grpc-java | spring-web
--- | ---
`@GrpcService` | `@Controller` / `@RestController`
`___ImplBase#<generatedMethod>` | `@RequestMapping` / `@PostMapping`
`*Interceptor` | `ServletFilter`
parameter `StreamObserver<T>` | returned `CompletableFuture/CompletableFuture<ResponseEntity<T>>`
returned `StreamObserver<T>` | no equivalent / listenable parameter `<T>` stream

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/service/exceptionhandling/GrpcExceptionAspect.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+ * 
+ * TODO...
+ *
+ * @author Andjelko (andjelko.perisic@gmail.com)
+ */
+@Slf4j
+@Aspect
+@Component
+@ConditionalOnBean(annotation = GrpcServiceAdvice.class)
+public class GrpcExceptionAspect {
```
[the remark this answers is not in the document]
Comment (anjeyy) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/service/exceptionhandling/GrpcExceptionAspect.java:52]: The Aspect is beeing executed for every Thread. If you call two times a GrpcService method and in your application throws respectively two times an exeception, the Aspect is beeing called twice, for every request. So of course if you have a share resource, you have to pay attention.

In my mind this was no problem, since for every client call we have a response and the response is beeing handled for every call (Thread) seperately.

This Aspect is working with Spring AOP component, which as i remember wraps a proxy around it, thats why springs aop is lightweigth in relation to the AspectJ AOP variant.

https://docs.spring.io/spring-framework/docs/5.3.2/reference/html/core.html#aop-api

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/service/exceptionhandling/GrpcExceptionAspect.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    private Object[] determineInstancedParameters(Method mappedMethod) {
+
+        Parameter[] parameters = mappedMethod.getParameters();
+        Object[] instancedParams = new Object[parameters.length];
+
+        for (int i = 0; i < parameters.length; i++) {
+            Class<?> parameterClass = convertToClass(parameters[i]);
+            if (parameterClass.equals(exception.getClass())) {
```
[the remark this answers is not in the document]
Comment (anjeyy) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/service/exceptionhandling/GrpcExceptionAspect.java:121]: You are right, i did conside this in `GrpcExceptionHandlerMethodResolver` but in the Aspect class i forgot to check if mapped Exception is superclass of thrown class.

I'll commit changes.

**Comment by anjeyy** [comment]
> You probably know this already, but I thought about adding this dictionary to the docs eventually.

Thank you for the comparison, just curios about the last one _returned `StreamObserver<T>`. Does this mean there is a continous connection established between client/server to _stream_ data? (via HTTP/2)

---


> Thanks for the link.
> Could you please also send me a link to the spring-web variant of `GrpcExceptionAspect`?

I am currently not aware of, if spring realizes their exception handling with an aspect. I did the Aspect since it was the most fortunate approach in my mind. Also the earlier mentioned [`ExceptionHandlerMethodResolver`](https://github.com/spring-projects/spring-framework/blob/master/spring-web/src/main/java/org/springframework/web/method/annotation/ExceptionHandlerMethodResolver.java), i am curios how it's beeing invoked by spring since this is obviously their class to handle exception mappings. I assume they use some kind of AOP, but not sure.


> I'm aware of that. But IMO this `asException()` is unnecessary. What else could a developer intend other than this status being sent to the client? If any exception were to be sent to the client as is, then I wouldn't mind, but the exceptions are effectively reduced to a `Status` message. So it should be possible to keep the user code free of these conversions/wrappers.
> In spring I prefer not no use the `ResponseEntity` wrappers unless I actually need them.
> 
> > How would the workflow look if the Advice is returning Status? I think the stream needs to be closed anyway, right? So instead of letting the developer have to think about it, the Aspect is closing the steam with the resulting error. Or do you think to leave it open to the developer?
> 
> No, the aspect would look and behave exactly the same, except for the check, if the exception handler returned a Status or an Exception. The Status would be converted to an exception and then be passed on.

Okay i see your point, i can change it in a bit. Should not be that much of a change.


> I completely agree to that (on the usage side that is).
> But there is one point that we needs to keep in mind. grpc also has streaming calls/is implicitly (potentially) async (as you can return from the controller method, without completing the StreamObserver) so it should be possible to just "quit" processing anywhere in the code in any thread by calling `StreamObserver#onError && return` and still have the exception handling kick in.
> 
> This is similar to default spring/spring-security, where you can (to a certain degree) handle exceptions outside of your controller implementation. IIRC the web exception handling also kicks in for methods that return unresolved `Future`s.

Whats the correct call to quit processing the stream operation? I thought it was enough just to call `StreamObserver<?>.onError(..)`, but as i am now understanding we need to take futher actions, to explicitly close the async request?

---

I'll turn next to JUnit Tests and various Test scenarios for mapping exceptions.

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/service/exceptionhandling/GrpcExceptionAspect.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (anjeyy) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/service/exceptionhandling/GrpcExceptionAspect.java]: done

**Comment by anjeyy** [comment]
@ST-DDT 
Im stuck on the Junit (Integration) Tests. I got it so far wired up and running with the AutoConfiguration, but i am unsure why a thrown Exception inside a test is not being caught from the aspect `GrpcServiceAdviceExceptionHandler`. (manually it has been working, everything i commited, i doubled checked at least manually)

If you could help me out getting the exception to be caught, i can then continue to add more test cases. `GrpcServiceAdviceTest` is the somewhat template class to get it running, with `GrpcServiceAdviceConfig` and `BaseExceptionAdviceAutoConfiguration`.

**Comment by ST-DDT** [comment]
The issue is probably that the proxy class delegates all (unchanged) methods to the original class, however it may not "unfinal"/delegate BindableService#bindService implementations, as that method returns method references to the bean being proxied.
If it removes the final modifier from the method, then it might break spring security annotation driven configuration.

**Comment by ST-DDT** [comment]
Updated my comment

**Comment by anjeyy** [comment]
I did research a bit, and found quite interesting food for thought.

- [stackoverflow](https://stackoverflow.com/a/50601574/11770752)
- referenced [github issue](https://github.com/grpc/grpc-java/issues/1552#issuecomment-219476882)
- [`TransmitStatusRuntimeExceptionInterceptor`](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/util/TransmitStatusRuntimeExceptionInterceptor.java) - maybe the thing we were looking for? 💭 

I'll give it a try and see if i can make something out of it. 🤔

**Comment by ST-DDT** [comment]
This looks pretty much like the interceptor I had in mind, maybe only lacking the exception handling for exceptions passed directly to `StreamObserver#onError`. (IIRC This can be achieved by wrapping the `ServerCall` instance with the exception handling logic).
I know this is a complex subject. So if you need support with anything, let me know and I will try to help you.

**Comment by anjeyy** [comment]
@ST-DDT 
i'll changed it to work with interceptors, if you could please have a look. 

Next step would be JUnit tests. And things from your side of course.

**Comment by anjeyy** [comment]
JUnit Tests have been added. Next thing up is updating the documentation.

Hopefully i did not mess up too bad, since i tried to adopt your given structure.

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdviceExceptionListener.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        } else if (mappedReturnType instanceof StatusRuntimeException) {
+            StatusRuntimeException statusException = (StatusRuntimeException) mappedReturnType;
+            result = statusException.getTrailers();
+        }
+        return (result == null) ? new Metadata() : result;
+    }
+
+    private void handleThrownExceptionByImplementation(Throwable throwable) {
+        serverCall.close(Status.INTERNAL.withCause(throwable)
+                .withDescription(throwable.getMessage()), new Metadata());
```
Comment (ST-DDT) [inline grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdviceExceptionListener.java:104]: This should log the stacktrace to the server logs. Also the description sent to the client should bebmote generic to avoid exposing any internal details.
E.g. `There was a server error trying to handle an exception.`

File: grpc-server-spring-boot-autoconfigure/src/main/resources/META-INF/spring.factories [hunk 4]
```diff
@@ -7,4 +7,5 @@ net.devh.boot.grpc.server.autoconfigure.GrpcServerAutoConfiguration,\
 net.devh.boot.grpc.server.autoconfigure.GrpcServerFactoryAutoConfiguration,\
 net.devh.boot.grpc.server.autoconfigure.GrpcServerSecurityAutoConfiguration,\
 net.devh.boot.grpc.server.autoconfigure.GrpcServerMetricAutoConfiguration,\
-net.devh.boot.grpc.server.autoconfigure.GrpcServerTraceAutoConfiguration
+net.devh.boot.grpc.server.autoconfigure.GrpcServerTraceAutoConfiguration,\
+net.devh.boot.grpc.server.autoconfigure.GrpcAdviceAutoConfiguration
```
Comment (ST-DDT) [inline grpc-server-spring-boot-autoconfigure/src/main/resources/META-INF/spring.factories:11]: Please retain the newline at EOF.

**Review by ST-DDT — COMMENTED** [review]
Looks good so far. I'll do a thorough review after x-mas.

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdviceExceptionListener.java (hunk 3 above)
Comment (ST-DDT) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdviceExceptionListener.java:104]: You could add the original exception as a surpressed exception. Just to give the developer a clue how to reproduce the error.

**Comment by ST-DDT** [comment]
FYI: The build/tests are failing randomly due to an async processing issue: https://github.com/yidongnan/grpc-spring-boot-starter/pull/471

**Comment by anjeyy** [comment]
> FYI: The build/tests are failing randomly due to an async processing issue: #471

Okay, I was wondering what was happening since I could not resolve the issue that caused the build to fail.

Anything I can do here, or just waiting #471 to be merged and then rebasing?

**Comment by ST-DDT** [comment]
> > FYI: The build/tests are failing randomly due to an async processing issue: #471
> 
> Okay, I was wondering what was happening since I could not resolve the issue that caused the build to fail.
> 
> Anything I can do here, or just waiting #471 to be merged and then rebasing?

Unfortunately, we can only wait for it to be merged.

File: tests/src/test/java/net/devh/boot/grpc/test/advice/GrpcMetdaDataUtils.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package net.devh.boot.grpc.test.advice;
+
+import io.grpc.Metadata;
+
+public class GrpcMetdaDataUtils {
```
Comment (ST-DDT) [inline tests/src/test/java/net/devh/boot/grpc/test/advice/GrpcMetdaDataUtils.java:22]: Is this spelled correctly?

**Review by ST-DDT — CHANGES_REQUESTED** [review]
I found some minor things, that should be changed, but nothing major.
Thanks for the nice documentation and your all your work.
Next week I will not be available, so I will try to merge it afterwards.

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdviceExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+ * Returned object is declared in {@link GrpcAdvice @GrpcAdvice} classes with annotated methods
+ * {@link GrpcExceptionHandler @GrpcExceptionHandler}.
+ * <p>
+ *
+ * @author Andjelko Perisic (andjelko.perisic@gmail.com)
+ * @see GrpcExceptionHandlerMethodResolver
+ * @see GrpcAdviceExceptionInterceptor
+ */
+@Slf4j
+@Aspect
```
[the remark this answers is not in the document]
Comment (anjeyy) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdviceExceptionHandler.java:45]: No, just forgot to remove it. Will do.

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdvice.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+ * <p>
+ * 
+ * @author Andjelko Perisic (andjelko.perisic@gmail.com)
+ * @see GrpcExceptionHandler
+ */
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Component
+public @interface GrpcAdvice {
```
[the remark this answers is not in the document]
Comment (anjeyy) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdvice.java:42]: You mean to be able to declare configuration classes with a GrpcAdvice `@Bean` ?
I think this should work by extending `@Retention`. I'll change it and extend the test.

Just want to point out, that the `@ControllerAdvice` has the same `RetentionPolicy` like `GrpcAdvice` _currently_.

File: docs/en/server/exception-handling.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+As you do not need `Metadata` in your response, just return your specified `Status`.
+
+### Overview of returnable types
+Here is a small overview of possible mapped return types with `@GrpcExceptionHandler` and if
+custom Metadata can be returned.
+
+| return type | Status | StatusException | StatusRuntimeException | Throwable |
+|-----------------|---------|----------|----------|---------|
+| custom Metadata | &cross; | &#10004; | &#10004; | &cross; |
```
[the remark this answers is not in the document]
Comment (anjeyy) [reply docs/en/server/exception-handling.md:95]: Yeah you're right, this is more readable and intuitive.

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdviceExceptionListener.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        this.serverCall = serverCall;
+        this.exceptionHandler = grpcAdviceExceptionHandler;
+    }
+
+    @Override
+    public void onHalfClose() {
+        try {
+            super.onHalfClose();
+        } catch (Exception exception) {
+            log.error("Exception caught during gRPC execution: ", exception);
```
[the remark this answers is not in the document]
Comment (anjeyy) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdviceExceptionListener.java:61]: I moved the logging to `GrpcAdviceExceptionHandler` with debug level.

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdviceExceptionListener.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+ * <b>Note:</b> In case of raised exceptions by implementation a {@link Status#INTERNAL} is returned in
+ * {@link #handleThrownExceptionByImplementation(Throwable)}.
+ *
+ * @param <ReqT> gRPC request type
+ * @param <RespT> gRPC response type
+ * @author Andjelko Perisic (andjelko.perisic@gmail.com)
+ * @see GrpcAdviceExceptionHandler
+ */
+@Slf4j
+public class GrpcAdviceExceptionListener<ReqT, RespT> extends SimpleForwardingServerCallListener<ReqT> {
```
[the remark this answers is not in the document]
Comment (anjeyy) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdviceExceptionListener.java:42]: The official documentation of `onHalfClose` is _The client completed all message sending. However, the call may still be cancelled._ So i was considering the streaming messages would be also handled properly, like in the tests and with manual testing. Or am i overlooking something?

In case to include (and override) the `onMessage`, which `ServerCall`to use?

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdviceExceptionInterceptor.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        this.grpcAdviceExceptionHandler = grpcAdviceExceptionHandler;
+    }
+
+    @Override
+    public <ReqT, RespT> Listener<ReqT> interceptCall(
+            ServerCall<ReqT, RespT> call,
+            Metadata headers,
+            ServerCallHandler<ReqT, RespT> next) {
+        Listener<ReqT> delegate = next.startCall(call, headers);
+        return new GrpcAdviceExceptionListener<>(delegate, call, grpcAdviceExceptionHandler);
```
[the remark this answers is not in the document]
Comment (anjeyy) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcAdviceExceptionInterceptor.java:49]: I had no thrown exception for streaming calls, i guess since the `server.close()` is called?

**Comment by ST-DDT** [comment]
FYI: #741 has been merged.

**Comment by anjeyy** [comment]
> I found some minor things, that should be changed, but nothing major.
> Thanks for the nice documentation and your all your work.
> Next week I will not be available, so I will try to merge it afterwards.

I changed the minor things, except for one thing that i could not clarify for myself. Replied in the comment.

Thank you for your kind words.

**Review by ST-DDT — CHANGES_REQUESTED** [review]
Sorry for the delay with the review, I was on vacation and didn't have access to my computer during that time.

AFAICT this PR does not cover exception handling during streaming calls and exceptions during responseObserver#onError. But that's not an issue, we can cover that in another/later PR and just focusing on having a solid base for the exception handling now.

Aside from a few spelling issues (I'm not a native English speaker, but some part feel strange to me) that I can change when merging this (or if you would like to review them, send them as a PR to you), there might be some places, where the code is too simple and searches for a "first match" instead of a "best match" such as `GrpcExceptionHandlerMethodResolver#extractExtendedThrowable`. But if we iron out this bit by bit, we should soon have a rock-solid basis for our error handling.

**Comment by anjeyy** [comment]
> Aside from a few spelling issues (I'm not a native English speaker, but some part feel strange to me) that I can change when merging this (or if you would like to review them, send them as a PR to you), [...]

Feel free to change the parts that may occur strange, since i am not a native englisch speaker myself. 😅 
<br>

> [...] there might be some places, where the code is too simple and searches for a "first match" instead of a "best match" such as `GrpcExceptionHandlerMethodResolver#extractExtendedThrowable`. But if we iron out this bit by bit, we should soon have a rock-solid basis for our error handling.

There are various checks when scanning the annotation in the beginning so that the _first match_ should be the _best match_, but may be i did overlook something. I tried to test the scenarios.

In any case, you see room for improvement, feel free to make changes inside this PR. 😃

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcExceptionHandlerMethodResolver.java [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+ * For an example how to make use of it, please have a look at {@link GrpcExceptionHandler @GrpcExceptionHandler}.
+ * <p>
+ * 
+ * @author Andjelko Perisic (andjelko.perisic@gmail.com)
+ * @see GrpcAdvice
+ * @see GrpcExceptionHandler
+ * @see GrpcAdviceExceptionHandler
+ */
+@Slf4j
+public class GrpcExceptionHandlerMethodResolver implements InitializingBean {
```
[the remark this answers is not in the document]
Comment (anjeyy) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/advice/GrpcExceptionHandlerMethodResolver.java:49]: I'm not sure if i understand your example correctly. May be if you could give me a more specific example, like in this case with the `IllegalArgumentException` and the Mapping that you define with `@GrpcExceptionHandler` i'll understand better.

---

A major part is happening before `extractExtendedThrowable` is being executed (filling the Exception Map).
More specific in method `GrpcExceptionHandlerMethodResolver#extractExceptions`: Annotated Exceptiontypes (in the annotation `@GrpcExceptionHandler`) are of a higher priority than the Exception types listed as method arguments.

 **But** a check is happening if the mapped Annotation Exception type is equal or a super type of any listed method Exception type, just to be sure when a exception is raised that the instance is passed in as a (correct - equal or subtype instance) parameter for the actual exception handling.

File: tests/src/test/java/net/devh/boot/grpc/test/advice/AbstractSimpleServerClientTest.java [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+        assertNotNull(this.channel, "channel");
+        assertNotNull(this.testServiceBlockingStub, "testServiceBlockingStub");
+        assertNotNull(this.testServiceFutureStub, "testServiceFutureStub");
+        assertNotNull(this.testServiceStub, "testServiceStub");
+    }
+
+    /**
+     * Test template call to check for every exception.
+     */
+    <E extends RuntimeException> void testGrpcCallAndVerifyMappedException(Status expectedStatus, Metadata metadata) {
```
[the remark this answers is not in the document]
Comment (anjeyy) [reply tests/src/test/java/net/devh/boot/grpc/test/advice/AbstractSimpleServerClientTest.java:73]: You're right.

**Comment by ST-DDT** [comment]
I started with the in depth review. It might take a few days to check for all corner cases.

**Review by ST-DDT — APPROVED** [review]
Sorry for the long review delay. This now looks good to me.
Thanks for your contribution!

**Comment by anjeyy** [comment]
Hi,

no need to excuse yourself, glad it worked out. And thank you for your improvements. 👍🏼
